### DISCUSSION TITLE
How do I set a fixed parameter range of MCMC?

### DISCUSSION THREAD

**Discussion by UltraX3D** [phoebe-project/phoebe2#724] [Questions]
Hello!
I set an initial parameter range, but the distribution of parameters I get after doing the MCMC is not within the range I set initially and is obviously not reasonable. So I would like to ask how to set a fixed parameter range, so that the parameters obtained after running MCMC are distributed within the parameter range I set initially.
Here's my script:

**b.add_distribution({'q': phoebe.gaussian_around(0.01),
                    'fillout_factor': phoebe.gaussian_around(0.1),
                    'incl@binary': phoebe.gaussian_around(7),
                     'teff@primary': phoebe.gaussian_around(12),
                    'teff@secondary': phoebe.gaussian_around(200),
                   'pblum@primary@lcw': phoebe.gaussian_around(2)},
                   distribution='ball_around_optimized_solution')

b.add_solver('sampler.emcee',
             init_from='ball_around_optimized_solution',
             compute='phoebe01',
             solver='emcee_solver',overwrite=True)

b.run_solver('emcee_solver', niters=1000, nwalkers=24, solution='emcee_sol',overwrite=True)**

**Comment by kecnry** [comment]
This can be achieved with uniform/boxcar priors, but note that you are then assuming this is known information in advance.  If placing priors to forbid wandering outside physical parameter space results in systematic residuals or the posteriors "running up against" the edges of the priors, then you'll likely need to add more freedom into the model so those residuals can be accounted for "correctly".

**Comment by UltraX3D** [comment]
I try it, but something is wrong,.It seems unable to set a priors. Some errors are listed below.
My code:

```
b.add_distribution({'q':` phoebe.uniform(0.5,0.7),
                    'fillout_factor': phoebe.uniform(0.1,0.25),
                    'incl@binary': phoebe.uniform(70,80),
                     'teff@primary': phoebe.uniform(6663,6687),
                    'teff@secondary': phoebe.uniform(6000,6500),
                   'pblum@primary@lcw': phoebe.uniform(11,14)},
                   distribution='ball_around_optimized_solution2')

b.add_solver('sampler.emcee', solver='emcee_solver',overwrite=True)
b.set_value('priors@emcee_solver',['ball_around_optimized_solution2'],overwrite=True)
b.set_value('init_from', ['ball_around_optimized_solution2'],overwrite=True)
b.set_value('compute', solver='emcee_solver', value='phoebe01')
b.run_solver('emcee_solver', niters=1000, nwalkers=24, solution='emcee_sol',overwrite=True)
```

Error:

ValueError                                Traceback (most recent call last)
~\AppData\Local\Temp\ipykernel_3632\3616204983.py in <module>
----> 1 b.run_solver('emcee_solver', niters=1000, nwalkers=24, solution='emcee_sol',overwrite=True)

~\anaconda3\lib\site-packages\phoebe\parameters\parameters.py in _send_if_client(self, *args, **kwargs)
    422 
    423         else:
--> 424             return fctn(self, *args, **kwargs)
    425 
    426     return _send_if_client

~\anaconda3\lib\site-packages\phoebe\frontend\bundle.py in run_solver(self, solver, solution, detach, return_changes, **kwargs)
  13647 
  13648         solver_class = getattr(_solverbackends, '{}Backend'.format(solver_ps.kind.title()))
> 13649         params = solver_class().run(self, solver_ps.solver, compute, solution=solution, **{k:v for k,v in kwargs.items() if k not in ['compute']})
  13650         metawargs = {'context': 'solution',
  13651                      'solver': solver_ps.solver,

~\anaconda3\lib\site-packages\phoebe\solverbackends\solverbackends.py in run(self, b, solver, compute, **kwargs)
    553 
    554         else:
--> 555             rpacketlists_per_worker = [self.run_worker(**packet)]
    556 
    557         logger.debug("rank:{}/{} calling _fill_solution".format(mpi.myrank, mpi.nprocs))

~\anaconda3\lib\site-packages\phoebe\solverbackends\solverbackends.py in run_worker(self, b, solver, compute, **kwargs)
   1537 
   1538                 logger.info("initializing {} walker positions".format(nwalkers))
-> 1539                 dc, params_uniqueids, p0 = b.sample_distribution_collection(distribution=init_from,
   1540                                                                             combine=init_from_combine,
   1541                                                                             include_constrained=False,

~\anaconda3\lib\site-packages\phoebe\frontend\bundle.py in sample_distribution_collection(self, twig, sample_size, as_quantity, set_value, keys, parameters, **kwargs)
   8949                 N = 1
   8950 
-> 8951         sampled_values = dc.sample(size=sample_size).T
   8952 
   8953         if require_checks or require_compute or (require_limits and not require_limits_success) or (require_priors and not require_priors_success):

~\anaconda3\lib\site-packages\phoebe\dependencies\distl\distl.py in sample(self, *args, **kwargs)
   4278         if kwargs.get('as_quantity', False):
   4279             array_kwargs['dtype'] = _np.dtype(object)
-> 4280         samples = _np.asarray([dist.sample(*args, seed=seeds, cache_sample=cache_sample, **sample_kwargs) for dist in self.dists], **array_kwargs).T
   4281 
   4282         if cache_sample:

~\anaconda3\lib\site-packages\phoebe\dependencies\distl\distl.py in <listcomp>(.0)
   4278         if kwargs.get('as_quantity', False):
   4279             array_kwargs['dtype'] = _np.dtype(object)
-> 4280         samples = _np.asarray([dist.sample(*args, seed=seeds, cache_sample=cache_sample, **sample_kwargs) for dist in self.dists], **array_kwargs).T
   4281 
   4282         if cache_sample:

~\anaconda3\lib\site-packages\phoebe\dependencies\distl\distl.py in sample(self, size, unit, as_quantity, wrap_at, seed, as_univariate, cache_sample)
   4843                 d.clear_cached_sample()
   4844 
-> 4845             return super(Composite, self).sample(size=size, unit=unit, as_quantity=as_quantity, wrap_at=wrap_at, seed=seed, cache_sample=cache_sample)
   4846         else:
   4847             # NOTE: even though in these cases we sample from the underlying children

~\anaconda3\lib\site-packages\phoebe\dependencies\distl\distl.py in sample(self, size, unit, as_quantity, wrap_at, seed, cache_sample)
   2751 
   2752         qs = _np.random.random(size=size)
-> 2753         sample = self.ppf(qs, wrap_at=False)
   2754         if cache_sample:
   2755             self._cached_sample = sample

~\anaconda3\lib\site-packages\phoebe\dependencies\distl\distl.py in ppf(self, q, unit, as_quantity, wrap_at)
   2260         """
   2261         try:
-> 2262             ppf = self.dist_constructor_object.ppf(q)
   2263         except AttributeError:
   2264             raise NotImplementedError("{} does not support ppf".format(self.__class__.__name__))

~\anaconda3\lib\site-packages\phoebe\dependencies\distl\distl.py in dist_constructor_object(self)
   1080         """
   1081         if self._dist_constructor_object_cache is None:
-> 1082             self._dist_constructor_object_cache = self.dist_constructor_func(*self.dist_constructor_args)
   1083 
   1084         return self._dist_constructor_object_cache

~\anaconda3\lib\site-packages\phoebe\dependencies\distl\distl.py in dist_constructor_args(self)
   4732             if cdf is None:
   4733                 # print("*** integrating to compute cdf over spline (this may be slow... we'll eventually cache this so it only needs to be done once until an attribute is changed)")
-> 4734                 spline = _interpolate.UnivariateSpline(x, pdf, k=1, s=0)
   4735                 def _spline_int_single(xi):
   4736                     return spline.integral(x[0], xi)

~\anaconda3\lib\site-packages\scipy\interpolate\_fitpack2.py in __init__(self, x, y, w, bbox, k, s, ext, check_finite)
    197                  ext=0, check_finite=False):
    198 
--> 199         x, y, w, bbox, self.ext = self.validate_input(x, y, w, bbox, k, s, ext,
    200                                                       check_finite)
    201 

~\anaconda3\lib\site-packages\scipy\interpolate\_fitpack2.py in validate_input(x, y, w, bbox, k, s, ext, check_finite)
    225         else:
    226             if not np.all(diff(x) > 0.0):
--> 227                 raise ValueError("x must be strictly increasing if s = 0")
    228         if x.size != y.size:
    229             raise ValueError("x and y should have a same length")

ValueError: x must be strictly increasing if s = 0

Can you give me some suggestion to solve this error?

**Reply by kecnry** [reply]
This is probably a bug in the underlying distributions package - can you let me know the version of scipy that you have installed?

That said, I wouldn't suggest using the same distribution for both `init_from` and `priors`, you'll probably want to define two different distributions where the `init_from` are gaussian around reasonable estimates and the `priors` are conservative uniform limits.  That alone should avoid this bug since the previous run was sampling correctly from the gaussian distributions, but if not, let me know and we'll try to prioritize reproducing and fixing the interpolation sampling issue.

**Comment by UltraX3D** [comment]
The version of scipy that I used is 1.9.1.
I used the gaussian distribution for 'init_from', uniform distribution for 'priors'. No more errors seem to be reported, but I'm not sure what happens if you use a different distribution. What happens to the posterior distribution of the parameters if we do that?

**Reply by kecnry** [reply]
The original positions of the walkers are drawn from the `init_from` distribution, and the `priors` are folded into the merit function, and so will _become_ part of the posterior distribution.  Sections 3 and 5.1 of [Conroy et al. 2020 ](http://phoebe-project.org/publications/2020Conroy+) explain how phoebe treats these distributions when calling emcee, and for details on emcee itself, see the [emcee docs and publications](https://emcee.readthedocs.io/en/stable/).

**Comment by UltraX3D** [comment]
Thank you so much, I got it.
